Fontlib

w1naenator opened the pull request:
1. Remove `libSceFont.sprx`, `libSceFontFt.sprx`, and `libSceFreeTypeOt.sprx` from the shadPS4 `sys_modules` directory (optional).

2. Dump the folders `/preinst/common/font` and `/system/common/font2` from your PS4.

3. Create a new folder (preferably inside the shadPS4 directory) named `sys_fonts`, and place the two dumped folders into it.

4. If the fonts are located outside the default shadPS4 `sys_fonts` directory, open `config.toml` in the shadPS4 directory and, under `[General]`, set `sysFontPath = "full\path\to\folder\where\font\and\font2\folders\located\"` to the path of the folder you created that contains the dumped font folders.

georgemoralis committed:
dummy fontlib libs

georgemoralis committed:
register font libs

georgemoralis committed:
typo fix

georgemoralis committed:
added font error file

georgemoralis committed:
added sceFontCharacterGetBidiLevel (from RE)

georgemoralis committed:
fixup

georgemoralis committed:
sceFontCharacterGetTextOrder , sceFontCharacterLooksFormatCharacters , sceFontCharacterLooksWhiteSpace RE

georgemoralis committed:
sceFontCharacterRefersTextBack,sceFontCharacterRefersTextNext,sceFontRenderSurfaceInit,sceFontRenderSurfaceSetScissor  RE

georgemoralis committed:
sceFontRenderSurfaceSetStyleFrame ,sceFontStyleFrameGetEffectSlant RE added

georgemoralis committed:
clang fix

georgemoralis committed:
sceFontStyleFrameGetEffectWeight

georgemoralis committed:
added sceFontStyleFrameGetScalePixel

georgemoralis committed:
Merge branch 'shadps4-emu:main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Update types.h

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
fixup merge

Commit:
Enhance font rendering and management system

- Integrated `stb_truetype` for advanced font rendering.
- Added support for external and system fonts with logging.
- Introduced new structures for glyphs, metrics, and kerning.
- Implemented functions for font state management and scaling.
- Improved glyph rendering, including subpixel and caching.
- Enhanced render surface initialization and scissor handling.
- Refactored stubbed functions with proper implementations.
- Added ABI compatibility checks for key structures.
- Improved logging, error handling, and code organization.
- Updated documentation and comments for better clarity.

georgemoralis committed:
Merge branch 'main' into fontlib

w1naenator committed:
Merge branch 'fontlib' into fontlib

w1naenator committed:
Merge branch 'main' into fontlib

georgemoralis commented on src/core/libraries/font/font.cpp:
```diff
@@ -995,37 +1723,14 @@ s32 PS4_SYSV_ABI sceFontStyleFrameGetResolutionDpi() {
 
 s32 PS4_SYSV_ABI sceFontStyleFrameGetScalePixel(OrbisFontStyleFrame* styleFrame, float* w,
```
I am pretty sure that this was decompiled 100% why it is nulled?

georgemoralis committed:
Merge branch 'main' into fontlib

Commit:
Fix for DC: font scale still inaccurate, cause unknown.

Refactor font library: streamline file handling, enhance layout caching, and improve scale computation.

Commit:
Merge branch 'fontlib' of https://github.com/w1naenator/shadPS4 into fontlib

Commit:
Remove unnecessary blank line in OrbisFontStyleFrame struct

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

Commit:
Loading font files from app0, some other fixes.

Commit:
restored accidentally deleted stubbs.

w1naenator committed:
Merge branch 'main' into fontlib

Commit:
Add system font path configuration and loading functionality

Config example:

[General]
...
sysFontPath = "/.../NotoSansJP-Regular.ttf"
...

Commit:
Merge branch 'fontlib' of https://github.com/w1naenator/shadPS4 into fontlib

Commit:
Handle missing system font by clearing font bytes and logging an error

Commit:
Remove unused <system_error> include from font.cpp

possible MacOS build crash reason.

Commit:
fontlib: refactor style frame API; config-driven system fonts

-Update style frame getters to take the frame explicitly:
--sceFontStyleFrameGetResolutionDpi(OrbisFontStyleFrame*, u32* h_dpi, u32* v_dpi)
--sceFontStyleFrameGetScalePixel(OrbisFontStyleFrame*, float* w, float* h)
--sceFontStyleFrameGetScalePoint(OrbisFontStyleFrame*, float* w, float* h)
-Ensure setters operate directly on the passed style frame (signatures unified):
--sceFontStyleFrameSetEffectSlant(OrbisFontStyleFrame*, float slantRatio)
--sceFontStyleFrameSetEffectWeight(OrbisFontStyleFrame*, float weightXScale, float weightYScale, u32 mode)
--sceFontStyleFrameSetResolutionDpi(OrbisFontStyleFrame*, u32 h_dpi, u32 v_dpi)
--sceFontStyleFrameSetScalePixel(OrbisFontStyleFrame*, float w, float h)
--sceFontStyleFrameSetScalePoint(OrbisFontStyleFrame*, float w, float h)
-Remove/streamline unused declarations from font.h to tighten the interface
-Call sites using style frame getters must pass an OrbisFontStyleFrame* now

Config: define system fonts path and default filename overrides

-Add [General].sysFontPath to set the base directory for system fonts
-Add [SystemFonts]:
--fallback = "SST-Roman.otf" (lowercase key) for the default face when no set is requested
--Per-font overrides using lowerCamelCase keys (start with fontSet...), e.g.:
---fontSetSstStdJapaneseJpArBold = "SSTJpPro-Bold.otf"
--Override values are filenames resolved under sysFontPath (paths are rejected and logged)
-Back-compat: still accept legacy SysFontPath and Fallback if present
-Logging: emit errors when sysFontPath is missing/invalid, fallback missing, or an override includes a path
-Configs should switch to sysFontPath and [SystemFonts].fallback with override keys

w1naenator committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

StevenMiller123 commented on src/common/config.cpp:
```diff
@@ -242,6 +246,44 @@ void setSysModulesPath(const std::filesystem::path& path) {
+
+void setSysFontPath(const std::filesystem::path& path) {
+    sys_font_path = path;
+}
```
I think it would be better to have the system font path as a shadPS4 directory instead of putting it in the config (to make it work like the shadPS4 sys_modules directory).

Telling users to throw all their PS4 font files in their shadPS4's font directory is a lot simpler than guiding them through editing their config.

w1naenator committed:
Merge branch 'main' into fontlib

Commit:
font.cpp

Added full glyph tracking/outline plumbing plus mutex-protected generated-glyph tracking and helpers (BuildTrueOutline, true-outline extraction, glyf detection, system-font cache/fallback logic, etc.).
Swapped OTF defaults for several JP/JP_AR sets (and CJK sets used by Death Stranding / Anywhere VR) to SSTAribStdB24-Regular.ttf or other TTFs so we load glyf faces.
Wrapped every LOG_* call in INFO/DEBUG pairs: INFO just announces the call, DEBUG lists each parameter on its own line with an extra trailing newline; warnings/errors remain for validation/fallback.
Implemented TTF detection for cache loading and fallback with logging.
Renamed glyph param struct/members, moved glyph handle struct into the header, and renamed the glyph API parameters/fields accordingly.
font.h

Added the new glyph outline/glyph handle definitions and renamed OrbisFontGenerateGlyphParams members (id, res0, form_options, mem, res1/2).
Updated function prototypes to use the renamed types and pointers (e.g., sceFontGenerateCharGlyph signature now matches the new names).
fontft.cpp

Reworked sceFontSelectLibraryFt/sceFontSelectRendererFt logging to follow the same INFO/DEBUG pattern (INFO announces that selection is requested, DEBUG dumps all params).

w1naenator committed:
Merge branch 'main' into fontlib

Commit:
clang

Commit:
Merge branch 'fontlib' of https://github.com/w1naenator/shadPS4 into fontlib

Commit:
Scaling and font state

Removed env‑driven debug scaling (ScaleMode, GetScaleMode, GetScaleBias, SHADPS4_FONT_SCALE_*).
Added FontState::is_otf_cff and storing of external vmetrics (ext_ascent, ext_descent, ext_lineGap).
Introduced ComputeScaleExtPoint, ComputeScaleExtPixel, IsEmProfileExternalFont and ComputeScaleExtForState to choose ascender‑ vs EM‑based scaling per font (system fonts + CFF → ascender; selected TTF profiles → EM).
Updated callers (sceFontGetCharGlyphMetrics, glyph cache, layout, renderer, clones, scale setters) to use ComputeScaleExtForState instead of ComputeScale / ComputeScaleExt.
System font handling

Added ReportSystemFaceRequest helper to centralize “attach system font or log why not” and avoid repeating the logic at each call site.
Slightly adjusted when system_requested is set and how fallback/system attach errors are reported.
Horizontal layout

sceFontGetHorizontalLayout now logs invalid parameter/handle and, when an external face is present, uses that face’s vmetrics and the per‑state scale to compute baselineOffset and lineAdvance.
(Fallback still uses the simple scale_h‑based approximation.)
Glyph rendering (horizontal)

Reworked sceFontRenderCharGlyphImageHorizontal:
Resolves effective pixel scale from the attached style frame + point/pixel API via ResolveStyleFrameScale.
Uses ComputeScaleExtForState to compute scale_y (and derived scale_x), rather than a single global scale.
Always computes metrics from stbtt_GetCodepointHMetrics + GetCodepointBitmapBoxSubpixel, with metrics->h_bearing_* and h_adv in sync with those scales.
Introduces a clearer origin/gravity model:
System fonts: (x,y) treated as raw top‑left.
Certain external TTFs (EM‑profile) : (x,y) treated as baseline.
Other external fonts (CFF/point etc.): (x,y) treated as line‑top with baseline offset.
Validates surfaces more strictly and returns NO_SUPPORT_SURFACE when buffer/size/bpp are invalid; supports only 1‑bpp and 4‑bpp.
Uses a local glyph_bitmap and a straightforward blit with scissor/clipping, instead of the older mix of cached bitmaps/PUA logging.
Removed PUA/placeholder glyph debug tracing and the more ad‑hoc baseline “adjusted_y” heuristic.
Logging clean‑up

Standardized LOG_INFO/LOG_DEBUG in sceFont* functions:
Info: usually just "called" or a short description (“scale pixel set requested”, etc.).
Debug: "parameters:\n"/"result:\n"/"template state:\n" followed by values on separate lines.
LOG_ERROR messages no longer repeat the function name (logger already prints it); they now say "invalid parameter", "invalid font handle", "no support glyph (face/scale)", "no support surface (buffer)", etc.
Added missing error logs before some early returns (allocation failures, page_count == 0, invalid parameters), and upgraded a few previous LOG_DEBUG “invalid params” to LOG_ERROR where they correspond to an error return.

w1naenator committed:
Merge branch 'main' into fontlib

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

w1naenator committed:
Merge branch 'main' into fontlib

Commit:
Add internal font handling structures and layout computation functions

This commit introduces a new header file `fontft_internal.h` containing
various structures and functions for managing font handles, layout metrics,
and rendering operations. Key additions include:

- Definitions for `FontHandleOpenInfo`, `FontHandleStyleStateTail`,
  and `FontHandleOpaqueNative` to encapsulate font state and properties.
- Layout-related structures such as `HorizontalLayoutBlocks`,
  `VerticalLayoutBlocks`, and their respective I/O structures for
  managing layout metrics and effects.
- Functions for computing horizontal and vertical layout blocks,
  along with utility functions for handling layout words in byte format.
- Stubs for library functions related to font management and rendering.

These changes aim to enhance the font rendering capabilities of the
shadPS4 Emulator Project, providing a robust foundation for future
font-related features.

Commit:
Add FreeType submodule for GitHub CI workflows

w1naenator committed:
Merge branch 'main' into fontlib

Commit:
Refactor font rendering and path resolution logic to fix Metaphor: ReFantazio glyphs cropping and anywhereVR  "thin" glyphs.

- Updated RenderCodepointToSurface and RenderCodepointToSurfaceWithScale functions to set result->transImage to nullptr instead of result->stage.
- Modified font definitions to use updated font file names for various language sets.
- Introduced ResolveSystemFontPathCandidate function to streamline font path resolution, allowing for better handling of font directories.
- Enhanced LoadFontFile function to check for font files in "font" and "font2" directories based on the parent directory name.
- Improved error logging for font path overrides in ResolveSystemFontPath function.

Commit:
clang

w1naenator committed:
Update fontft_internal.cpp

typo fix

Commit:
Fix: fypo

w1naenator committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

georgemoralis committed:
create sys_fonts in users folder and instructions.txt of what to put inside

georgemoralis committed:
cut version of emulator settings to support custom sysfonts folder

georgemoralis committed:
load settings

georgemoralis committed:
cmakelist

georgemoralis committed:
using new saving system but i can't understand the fallback mechanism

georgemoralis committed:
removed lle versions

Commit:
Enhance error logging in font metrics functions and update comments for clarity

Commit:
Refactor font library functions and improve thread safety

- Updated function signatures in font.h to include the OrbisFontLib parameter for cache clearing and outline buffer management.
- Introduced a mutex in font_internal.cpp to ensure thread safety by protecting access to global font state and library state.
- Added a RemoveState function to safely remove font states.
- Enhanced memory allocation functions in fontft_internal.cpp to use Core::ExecuteGuest for better integration with the guest environment.
- Simplified font loading logic in LibraryOpenFontMemoryStub to improve readability and maintainability.
- Removed redundant code related to file opening and memory allocation, streamlining the font loading process.
This potentially fixes emulator crashes in games that use parallel font initialization, such as Dragon’s Crown.

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

Commit:
Merge branch 'fontlib' of https://github.com/w1naenator/shadPS4 into fontlib

Commit:
Add system font path and override config support

Introduce new config options for system font directory, fallback font name, and per-font overrides. Update config load/save logic to handle a [SystemFonts] TOML section, supporting both fallback and individual font overrides. Improve user instructions for custom font setup and clarify related code comments. These changes enhance flexibility and user experience for system font configuration.

w1naenator committed:
Merge branch 'main' into fontfixes

w1naenator committed:
Merge branch 'fontfixes' into fontlib

Commit:
clang

Commit:
Add system font path management functions and refactor font directory retrieval

georgemoralis commented on src/emulator.cpp:
```diff
@@ -531,11 +529,7 @@ void Emulator::LoadSystemModules(const std::string& game_serial) {
          {"libSceJson2.sprx", nullptr},
          {"libSceLibcInternal.sprx", &Libraries::LibcInternal::RegisterLib},
-         {"libSceCesCs.sprx", nullptr},
-         {"libSceAudiodec.sprx", nullptr},
```
libSceAudiodec shouldn't be removed

w1naenator replied:
ups

Commit:
Restored support for loading the libSceAudiodec system module

w1naenator committed:
Merge branch 'main' into fontlib

w1naenator committed:
Merge branch 'main' into fontlib

w1naenator committed:
Merge branch 'main' into fontlib

w1naenator committed:
Merge branch 'main' into fontlib

w1naenator committed:
Merge branch 'main' into fontlib

Commit:
Refactor font path handling and remove unused functions

Commit:
Fix include order for emulator settings and key manager (Clang)

Commit:
Remove unused emulator settings files and clean up includes in main and config files

Commit:
Remove unused instructions file creation from user paths initialization

Commit:
Consolidate internal font/ft internals

Move internal font structs, aliases, and helpers out of font.cpp into shared internal headers/sources.
Promote shared ABI-facing types (including FontHandleOpaqueNative and layout/system-use structs) into font internal headers and remove function-local duplicates.
Move FreeType driver/renderer table construction into fontft_internal.cpp and wire fontft.cpp to internal getters.
Ensure internal.cpp implementations are declared in headers; deduplicate declarations and clean includes.

w1naenator committed:
Merge branch 'main' into fontlib

w1naenator committed:
Merge branch 'main' into fontlib

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

Commit:
Refactor memory allocation and deallocation calls to remove Core::ExecuteGuest wrapper

Commit:
clang

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

georgemoralis committed:
Merge branch 'main' into fontlib

w1naenator committed:
Merge branch 'main' into fontlib

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

Commit:
Add conditional linking for freetype library in shadps4 target

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

w1naenator committed:
Merge branch 'main' into fontlib

Commit:
Add new fixed-point arithmetic functions for font rendering

This commit introduces several new functions to handle fixed-point arithmetic operations, specifically for converting and scaling values related to font rendering. The added functions include:
- FixedMulUnitsToF26Dot6
- Cvttss2siCompat
- RoundMul16x16ToS32
- TruncFixed16x16ToInt
- RoundFixedMulValueScaleToS32
- TruncMulUnitsToS64
- RoundFixedMul16x16ToS64

These functions will enhance the precision and performance of font rendering calculations.

w1naenator committed:
Merge branch 'main' into fontlib

w1naenator committed:
Merge branch 'main' into fontlib

Commit:
Merge remote-tracking branch 'origin/main' into fontlib

# Conflicts:
#	CMakeLists.txt

w1naenator committed:
Merge branch 'main' into fontlib

Commit:
Merge remote-tracking branch 'origin/main' into fontlib

w1naenator committed:
Merge branch 'main' into fontlib

Commit:
Merge branch 'fontlib' of https://github.com/w1naenator/shadPS4 into fontlib

# Conflicts:
#	src/core/libraries/libs.cpp

w1naenator committed:
Merge branch 'main' into fontlib

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

w1naenator committed:
Merge branch 'shadps4-emu:main' into fontlib

Commit:
Refactor: Replace Config usage with EmulatorSettings for font directory retrieval

w1naenator committed:
Merge branch 'main' into fontlib

Commit:
Sysmodule: Stop forcing HLE for font modules

Commit:
Add support for built-in font handling and fallback mechanisms

- Introduced a new structure `BuiltinFontBlob` to manage built-in font data.
- Implemented decompression functions for loading built-in fonts from compressed data.
- Added functions to resolve and load built-in fonts, including `LoadBuiltinFontBytesShared` and `IsBuiltinFontPath`.
- Enhanced `AttachSystemFont` to utilize built-in fonts as fallbacks when system fonts are unavailable.
- Implemented `AddBuiltinFallbackFaces` to add necessary fallback fonts based on font set types.
- Updated font selection logic to include Thai and symbol fonts in `ResolveSysFontCodepoint`.
- Modified `LibraryOpenFontMemoryStub` to handle loading built-in fonts directly from shared data.
- Added notifications for users when system fonts are not found, indicating fallback usage.
- Updated header files to declare new structures and functions related to built-in font handling.

w1naenator committed:
Merge branch 'main' into fontlib

StevenMiller123 reviewed (approved):
Assuming this still works at least as well as it did when I last tested, I think this is fine to merge.
